src/vm/types/program.zig

```diff
@@ -33,32 +33,116 @@ pub const HintRange = struct {
     length: usize,
 };
 
+pub const HintsRanges = union(enum) {
+    const Self = @This();
+
+    Extensive: std.HashMap(
+        Relocatable,
+        HintRange,
+        std.hash_map.AutoContext(Relocatable),
+        std.hash_map.default_max_load_percentage,
+    ),
+    NonExtensive: std.ArrayList(HintRange),
+
+    pub fn isExtensive(self: *Self) bool {
+        return switch (self.*) {
+            .Extensive => true,
+            inline else => false,
+        };
+    }
+
+    pub fn count(self: *Self) usize {
+        return switch (self.*) {
+            .Extensive => self.Extensive.count(),
+            .NonExtensive => self.NonExtensive.items.len,
+        };
+    }
+
+    pub fn deinit(self: *Self) void {
+        switch (self.*) {
+            .Extensive => self.Extensive.deinit(),
+            .NonExtensive => self.NonExtensive.deinit(),
+        }
+    }
+};
+
+const Hints = std.ArrayList(HintParams);
+
 /// Represents a collection of hints.
 ///
 /// This structure contains a list of `HintParams` and a map of `HintRange` corresponding to a `Relocatable`.
 pub const HintsCollection = struct {
     const Self = @This();
     /// List of HintParams.
-    hints: std.ArrayList(HintParams),
+    hints: Hints,
     /// Map of Relocatable to HintRange.
-    hints_ranges: std.HashMap(
-        Relocatable,
-        HintRange,
-        std.hash_map.AutoContext(Relocatable),
-        std.hash_map.default_max_load_percentage,
-    ),
+    hints_ranges: HintsRanges,
 
     /// Initializes a new HintsCollection with default values.
     ///
     /// # Params:
     ///   - `allocator`: The allocator used to initialize the collection.
+    pub fn init(allocator: Allocator, hints: std.AutoHashMap(usize, []const HintParams), program_length: usize, extensive_hints: bool) !Self {
+        var max_hint_pc = @as(usize, @intCast(0));
+        var total_hints_len = @as(usize, @intCast(0));
+        var it = hints.iterator();
+        while (it.next()) |kv| {
+            max_hint_pc = @max(max_hint_pc, kv.key_ptr.*);
+            total_hints_len += kv.value_ptr.len;
+        }
+
+        if (max_hint_pc == 0 or total_hints_len == 0) {
+            return Self.initDefault(allocator);
+        }
+
+        if (max_hint_pc >= program_length) {
+            return ProgramError.InvalidHintPc;
+        }
+
+        var hints_values = try std.ArrayList(HintParams).initCapacity(allocator, total_hints_len);
+        var hints_ranges_non_ext: ?std.ArrayList(HintRange) = null;
+        var hints_ranges_ext: ?std.AutoHashMap(Relocatable, HintRange) = null;
+
+        if (extensive_hints) {
+            hints_ranges_ext = std.AutoHashMap(Relocatable, HintRange).init(allocator);
+        } else {
+            hints_ranges_non_ext = std.ArrayList(HintRange).init(allocator);
+        }
```

Comment:
I like a lot the way you define `HintsRanges` union(enum). Seeing your implementation, I said to myself that this whole block of code would go well in an `init` method of `HintsRanges` with your `extensive_hints` flag as an argument. This way here, rather than init either an `AutoHashMap` or an `ArrayList`, we would instead directly initialize an instance of `HintsRanges` what do you think?

Reply:
Great suggestion!